packages/gatsby-source-contentful/src/extend-node-type.js

```diff
@@ -547,18 +546,83 @@ const fluidNodeType = ({ name, getTracedSVG }) => {
   }
 }
 
-exports.extendNodeType = ({ type, store }) => {
+exports.extendNodeType = ({ type, store, cache }) => {
   if (type.name.match(/contentful.*RichTextNode/)) {
     return {
       nodeType: {
         type: GraphQLString,
-        deprecationReason: `This field is deprecated, please use 'json' instead.`,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
       },
       json: {
-        type: GraphQLJSON,
-        resolve: (source, fieldArgs) => {
-          const contentJSON = JSON.parse(source.internal.content)
-          return contentJSON
+        type: GraphQLString,
+        deprecationReason: `This field is deprecated, please use 'raw' instead. @todo add link to migration steps.`,
+      },
+      references: {
+        type: [`ContentfulReference`],
+        async resolve(source, args, context, info) {
+          const parent = await context.nodeModel.findRootNodeAncestor(source)
+
+          const rawReferences = { Entry: new Set(), Asset: new Set() }
+
+          // Locate all Contentful Links within the rich text data
+          const traverse = obj => {
+            for (let k in obj) {
+              const value = obj[k]
+              if (value && value.sys && value.sys.type === `Link`) {
+                rawReferences[value.sys.linkType].add(value.sys.contentful_id)
+              } else if (value && typeof value === `object`) {
+                traverse(value)
+              }
+            }
+          }
+
+          traverse(JSON.parse(source.raw))
+
+          if (!rawReferences.Entry.size && !rawReferences.Asset.size) {
+            return []
+          }
+
+          const rawEntries = Array.from(rawReferences.Entry)
+          const rawAssets = Array.from(rawReferences.Asset)
+
+          // Query for referenced nodes
+          const resultEntries = await context.nodeModel.runQuery({
+            query: {
+              filter: {
+                contentful_id: { in: rawEntries },
+              },
+            },
+            type: `ContentfulEntry`,
+          })
+
+          const resultAssets = await context.nodeModel.runQuery({
+            query: {
+              filter: {
+                contentful_id: { in: rawAssets },
+              },
+            },
+            type: `ContentfulAsset`,
+          })
+
+          // Localize results
+          const nodeLocale = parent.node_locale
+
+          const findForLocaleWithFallback = (nodeList = [], referenceId) =>
+            nodeList.find(
+              result =>
+                result.contentful_id === referenceId &&
+                result.node_locale === nodeLocale
+            )
+
+          const localizedEntryReferences = rawEntries.map(referenceId =>
+            findForLocaleWithFallback(resultEntries, referenceId)
```

Comment:
If an RTE contains a single link to a non-existent entry (an entry that was deleted after being linked), the call to `runQuery` will return null, which will in turn cause this call to throw an error:

```
There was an error in your GraphQL query:

Cannot read property 'find' of null

  20 |           }
  21 |         }
  22 |       }
  23 |     }
  24 |   }
  25 | }
  26 |
  27 | fragment RichTextFragment on ContentfulRichText {
  28 |   __typename
  29 |   raw
> 30 |   references {
```


The default value for `nodeList` is only used if the parameter is left undefined.

```suggestion
            findForLocaleWithFallback(resultEntries || [], referenceId)
```

Reply:
Good catch, thx!